### PR TITLE
ENH: inline yt.funcs.is_sequence

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

This is in support to YTEP 0038, and more broadly in support to better typing in yt.
It is a follow up to #2893, in which I renamed `has_len` into `is_sequence` based on a misconception
I was having about the word ["sequence" in Python](https://docs.python.org/dev/glossary.html#term-sequence).

I encourage reviewers to read this short page https://docs.python.org/3/library/collections.abc.html

So, the main reason this function needs to be discouraged is that its name is misleading. It is _not_ equivalent to the following implementation
```python
from collections.abc import Sequence
def is_sequence(obj):
    # check that `obj` has __getitem__, __len__, __reversed__ ...
    return isinstance(obj, Sequence)
```
On the other hand, it _is_ equivalent to
```python
from collections.abc import Sized
def is_sequence(obj):
    # check that obj has __len__
    return isinstance(obj, Sized)
```
and because this implementation is try-free as well as simpler, I replaced the current implem with it. 

Now, since it is apparent that it can now be inlined at the same cost (one import), it seems reasonable to drop its usage internally, but we can still keep it for now, to avoid breaking downstream applications.

In many places where it is used, checking that a given object knows the `__len__` protocol isn't exactly the appropriate thing to do. Sometimes we really want to check that objets are instances of `collections.abc.Sequence` or `collections.abc.Collection`. However, to avoid breaking anything by changing a lot of code at once, I stopped at the inlining step, except in one or two obvious cases, which I will highlight here with comments.

It is worth noting that `np.ndarray` objects are _not_ considered as `collections.abc.Sequence`, because they don't implement the `__reversed__` protocol. See https://docs.python.org/dev/glossary.html#term-sequence for more detail

## PR Checklist

- [N/A] New features are documented, with docstrings and narrative docs
- [N/A] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/data_objects/selection_objects/data_selection_objects.py [hunk 1]
```diff
@@ -594,7 +595,7 @@ def to_frb(self, width, resolution, center=None, height=None, periodic=False):
             )
 
             validate_width_tuple(width)
-            if is_sequence(resolution):
+            if isinstance(resolution, Iterable):
```
Comment (neutrinoceros) [inline yt/data_objects/selection_objects/data_selection_objects.py:598]: This is one of the few cases that I corrected the logic

File: yt/data_objects/time_series.py [hunk 2]
```diff
@@ -169,7 +170,7 @@ def __init__(
     ):
         # This is needed to properly set _pre_outputs for Simulation subclasses.
         self._mixed_dataset_types = mixed_dataset_types
-        if is_sequence(outputs) and not isinstance(outputs, str):
+        if isinstance(outputs, Sequence) and not isinstance(outputs, str):
```
Comment (neutrinoceros) [inline yt/data_objects/time_series.py:173]: Here too

File: yt/visualization/plot_container.py [hunk 3]
```diff
@@ -253,7 +254,11 @@ def __init__(self, data_source, figure_size, fontsize):
         self.data_source = data_source
         self.ds = data_source.ds
         self.ts = self._initialize_dataset(self.ds)
-        if is_sequence(figure_size):
+        if isinstance(figure_size, Sized):
+            if len(figure_size) != 2:
```
Comment (neutrinoceros) [inline yt/visualization/plot_container.py:258]: I added a sensible check here

**Comment by matthewturk** [comment]
Can you remind me if this has or had any impact on our usage of `ensure_list` to make sure that field lists are in fact lists of fields?

**Comment by neutrinoceros** [comment]
@matthewturk `ensure_list` was removed in #2893
The present PR is designed to have _no_ impact on anything, besides the two (extremely minor) corrections highlighted above.
It is only about avoiding to entertain an incorrect notion of what "sequences" mean in Python, and make room to later use `is instance(..., Sequence)` where relevant.

**Comment by neutrinoceros** [comment]
The GH workflow failures seem to be due to a change upstream on MacOs + home-brew + proj + gets + Python 3.9.7... So fixing this is out of scope here. However, the failures on Jenkins are real, so I'll make this a draft for now.

**Comment by neutrinoceros** [comment]
This allowed me to discover a bug in unyt https://github.com/yt-project/unyt/issues/201
though this problem is blocking for the current PR so I'll _keep_ it as a draft until it is resolved upstream.

**Comment by neutrinoceros** [comment]
Turns out the bug isn't in unyt, but in numpy. I think this PR is doomed because numpy arrays have weird behaviours and don't seem to play well with `collections.abc`. Down the line these problems completely break what I was trying to do here

See https://github.com/numpy/numpy/issues/19833, and a related, long standing issue https://github.com/numpy/numpy/issues/2776

I'll open another PR to try and reflect this knowledge better in our code base.

**Pull request closed without merging**